### DISCUSSION TITLE
Request: provide Windows ARM64 (and Linux ARM64) Python wheels on PyPI

### DISCUSSION THREAD

**Discussion by oscarbg** [halide/Halide#8516] [Ideas]
Hi,
congrats on https://pypi.org/project/halide/19.0.0/ release..
can provide a Win ARM64 build? ideally also a Linux ARM64 build..
for example for Python 3.13:
MacOS we get:
>halide-19.0.0-cp313-cp313-macosx_11_0_x86_64.whl
>halide-19.0.0-cp313-cp313-macosx_11_0_arm64.whl

but for Windows only x64:

>halide-19.0.0-cp313-cp313-win_amd64.whl 

would be nice to have:

>halide-19.0.0-cp313-cp313-win_arm64.whl 

similar in Linux..

thanks..

**Comment by alexreinking** [comment]
We would like to provide more binaries, and in fact we tried, but we ran into issues cross-compiling LLVM for ARM. The free GitHub Actions runners are simply too slow to go through qemu. Even worse, the manylinux containers that come with cross compiling toolchains are quite broken. Or at least they were back in August when I tried them. See for example, https://github.com/dockcross/dockcross/issues/849

We hope being on pip will attract more users that can help us fix this!

**Reply by alexreinking** [reply]
As for Windows on ARM, we currently have no buildbots/testing for that platform at all. I'm not even sure it builds there! We'd welcome contributions in this direction: anything from build instructions and/or patches to hardware donations.

**Reply by oscarbg** [reply]
just an user request.. no idea.. but seems coming soon are native ARM64 public Github action runners (Windows ARM64, and Linux ARM64) https://github.blog/news-insights/product-news/arm64-on-github-actions-powering-faster-more-efficient-build-systems/..
it's expected to help in this case by avoiding cross-compiling and Qemu entirely?

**Reply by alexreinking** [reply]
Unfortunately, those runners are considered "larger runners" and are not free, even for public, open-source repositories, like Halide.

**Comment by alexreinking** [comment]
This is looking much more possible! At least, we now have LLVM wheels for ARM64 and ARM32. There are outstanding build issues with ARM32 Python, but I'm working on those too.
